File: 953.verifying-an-alien-dictionary.cpp
/*
 * @lc app=leetcode id=953 lang=cpp
 *
 * [953] Verifying an Alien Dictionary
 */
#include <algorithm>
#include <string>
#include <vector>

using std::is_sorted;
using std::string;
using std::vector;

// @lc code=start
class Solution {
public:
    // T: O(s + n * l), s := len(order) (alphabet size), n := len(words), l := avg word length
    // S: O(s)
    bool isAlienSorted(vector<string>& words, string order) {
        int mapping[26];
        // Map order to normal order
        for (int i = 0; i < 26; ++i) {
            mapping[order[i] - 'a'] = i;
        }

        for (string &w : words) {
            for (char &c : w) {
                c = mapping[c - 'a'];   // changes strings inplace
            }
        }
        return is_sorted(words.begin(), words.end());
    }
};
// @lc code=end

File: 953.verifying-an-alien-dictionary.py
#
# @lc app=leetcode id=953 lang=python3
#
# [953] Verifying an Alien Dictionary
#
from typing import List

# @lc code=start
class Solution:
    def isAlienSorted(self, words: List[str], order: str) -> bool:
        mapping = {c: i for i, c in enumerate(order)}

        n = len(words)
        # new_words = [""] * n

        # Initialized to (0, ) to compare with 
        # the first rearranged word (i == 0)
        pre = (0, )
        for i in range(n):
            # Tuple of character indexes is enough for checking,
            # no need to convert to a string
            w = tuple([mapping[c] for c in list(words[i])])

            # new_words[i] = w
            if pre > w:
                return False
            pre = w
        return True

# @lc code=end

